File: Matlab Version (Deprecated)/App and Required Functions/DualV2.m
% Dual mode CCC by reversing the stationary while keep mobile phase
% written by Hoon Choi
% Loading New SP into a column with reverse flow

%Xtot: mobile phase concentration from classic elution
%Ytot: stationary phase concentration from classic elution
%Col_out: Column outlet proflies for MP and SP respectively
%X: Concentration in MP during dual mode
%Y: Concentration in SP during dual mode


function [Vspan Cout, X, Y] = DualV2(KD, Vc, Sf, F, Vdm, Xcm, Ycm)

P = Sf/(1-Sf);
Vm = Vc*(1-Sf);
Vs = Vc*Sf;
kA = 1./(1+P.*KD); %Retention Factor!



[Ncup el_time comp] = size(Xcm);   %Ccup profile (Ncup,time,species)

vmcup = Vm/Ncup;
vscup = Vs/Ncup;

tCM = el_time;
VCMspan = vmcup.*linspace(1,tCM,tCM); %previous CM elution timestep

dt = vscup/F;
Tau = round(Vdm/vscup);  %elution timestep for DM


X = zeros(Ncup, Tau, comp); % MP concentration
Y = zeros(Ncup, Tau,  comp); %SP Concentration g/L


%take previous profiles from classic elution to set initial boundary
Xin = zeros(Ncup,1,comp);
Yin = zeros(Ncup,1,comp);



for h = 1:comp

    Xin(:,1,h) = Xcm(:,el_time,h);
    Yin(:,1,h) = Ycm(:,el_time,h);       %KD(comp).*Xin(:,1,comp);
    
end


for j = 1:comp  %comp
    
    t=1; %initial condition at t = 1
    
    X(Ncup,1,j) = kA(j)*Xin(Ncup,1,j);
    Y(Ncup,1,j) = KD(j)*X(Ncup,1,j);

    for i = Ncup:-1:2  %calculate from Ncup to 1st cell
        
        X(i-1,t,j) = kA(j)*(Xin(i-1,1,j)+P*Yin(i,1,j));
        Y(i-1,t,j) = KD(j)*X(i-1,t,j);
               
    end
    
              
     for t = 2:Tau % SP reverse elution (dT = Vscup) 
                
         
         % Boundary condition first for Ncup cell
         X(Ncup,t,j) = kA(j)*X(Ncup,t-1,j);
         Y(Ncup,t,j) = KD(j)*X(Ncup,t,j);
         
         %SP reverse elution from [Ncup-1] to 1st cell
         for i = Ncup:-1:2
                
             X(i-1,t,j) = kA(j)*(X(i-1,t-1,j)+P*Y(i,t-1,j));
             Y(i-1,t,j) = KD(j)*X(i-1,t,j);

         end
         
     end

     %Combine concentration profiles into Xtot & Ytot from CM to EECCC
          Xtot = [Xcm  X];
          Ytot = [Ycm  Y];
     

          %make elution time and volume matrix for EECCC 
          tspan = linspace(1, Tau, Tau);
          Vdm = vscup.*tspan;
          Vdm = VCMspan(end) + Vdm;

          Vspan = [VCMspan  Vdm];  %total elution volume
          
            %save outlet concentration 
        CoutDM = [];
         for h = 1:comp
            
                CoutCM (h,:) = Xcm(Ncup,:,h);
            
               for t = 1:Tau
                    
                  CoutDM (h,t) = Y(1,t,h);
               
               end


         end

         Cout = [CoutCM  CoutDM];
              
%           
%           for i = 1:n
%               
%               M(i,:) = Y(1,:,i);   % column outlet is 1st cell in dual mode
%               M(i+n,:)= X(1,:,i);  % Column outlet is 1st cell
%                                          
%           end
%           
%           Col_out = [Vspan; M]';  % [Vol yi  xi] column outlet profiles
%                 


end

          
     
  

File: models/DualV2.py
from numba import jit, prange
import numpy as np

@jit(nopython=True, parallel=True)
def DualV2(KD, Vc, Sf, F, Vdm, Xcm, Ycm):
    """
    Parameters:
    KD: Distribution coefficient ([Conc_SP]eq/[Conc_MP]eq)
    Vc: Total column volume
    Sf: Stationary factor (volume ratio of [V_SP]/[V_MP])
    F: Flow rate
    Vdm: Elution volume for dual mode
    Xcm: MP profiles from classic elution (Ncup x el_time x comp)
    Ycm: SP profiles from classic elution (Ncup x el_time x comp)

    Returns:
    Vspan: Total elution volume
    Cout: Column outlet profiles for MP and SP
    X: Concentration in MP during dual mode
    Y: Concentration in SP during dual mode
    """
    P = Sf / (1 - Sf)
    Vm = Vc * (1 - Sf)
    Vs = Vc * Sf
    kA = 1 / (1 + P * KD)  # Retention factor

    Ncup, el_time, comp = Xcm.shape  # Cup profile dimensions
    vmcup = Vm / Ncup
    vscup = Vs / Ncup

    tCM = el_time
    VCMspan = np.zeros(tCM)
    for i in range(tCM):
        VCMspan[i] = vmcup * (i + 1)  # Previous CM elution timestep

    dt = vscup / F
    Tau = int(round(Vdm / vscup))  # Elution timestep for dual mode

    X = np.zeros((Ncup, Tau, comp))  # MP concentration
    Y = np.zeros((Ncup, Tau, comp))  # SP concentration

    # Process each component in parallel
    for j in prange(comp):
        # Initial condition at t = 0
        X[Ncup - 1, 0, j] = kA[j] * Xcm[Ncup - 1, el_time - 1, j]
        Y[Ncup - 1, 0, j] = KD[j] * X[Ncup - 1, 0, j]

        # Calculate from Ncup to 1st cell for t=0
        for i in range(Ncup - 1, 0, -1):
            X[i - 1, 0, j] = kA[j] * (Xcm[i - 1, el_time - 1, j] + P * Ycm[i, el_time - 1, j])
            Y[i - 1, 0, j] = KD[j] * X[i - 1, 0, j]

        # SP reverse elution (dT = Vscup)
        for t in range(1, Tau):
            # Boundary condition for Ncup cell
            X[Ncup - 1, t, j] = kA[j] * X[Ncup - 1, t - 1, j]
            Y[Ncup - 1, t, j] = KD[j] * X[Ncup - 1, t, j]

            # SP reverse elution from [Ncup-1] to 1st cell
            for i in range(Ncup - 1, 0, -1):
                X[i - 1, t, j] = kA[j] * (X[i - 1, t - 1, j] + P * Y[i, t - 1, j])
                Y[i - 1, t, j] = KD[j] * X[i - 1, t, j]

    # Combine concentration profiles into Xtot & Ytot from CM to dual mode
    Xtot = np.zeros((Ncup, el_time + Tau, comp))
    Ytot = np.zeros((Ncup, el_time + Tau, comp))

    # Copy CM profiles
    for i in range(Ncup):
        for t in range(el_time):
            for j in range(comp):
                Xtot[i, t, j] = Xcm[i, t, j]
                Ytot[i, t, j] = Ycm[i, t, j]

    # Copy dual mode profiles
    for i in range(Ncup):
        for t in range(Tau):
            for j in range(comp):
                Xtot[i, el_time + t, j] = X[i, t, j]
                Ytot[i, el_time + t, j] = Y[i, t, j]

    # Make elution time and volume matrix for dual mode
    Vdm_span = np.zeros(Tau)
    for i in range(Tau):
        Vdm_span[i] = vscup * (i + 1) + VCMspan[-1]

    # Total elution volume
    Vspan = np.zeros(el_time + Tau)
    for i in range(el_time):
        Vspan[i] = VCMspan[i]
    for i in range(Tau):
        Vspan[el_time + i] = Vdm_span[i]

    # Save outlet concentration
    Cout = np.zeros((comp, el_time + Tau))
    for h in range(comp):
        # CM outlet
        for t in range(el_time):
            Cout[h, t] = Xcm[Ncup - 1, t, h]
        # DM outlet
        for t in range(Tau):
            Cout[h, el_time + t] = Y[0, t, h]

    return Vspan, Cout, X, Y
